### ISSUE TITLE
[Vertical Taskbar for Windows 11] Use a grid for tray icons

### ISSUE THREAD

**Issue by m417z** [ramensoftware/windhawk-mods#1068]
### Discussed in https://github.com/ramensoftware/windhawk-mods/discussions/1067

<div type='discussions-op-text'>

<sup>Originally posted by **WegnerDan** October 12, 2024</sup>
I really like that I can finally have a vertical taskbar in win11! 
But the tray icon area is really large. On Win10 and before, I used to have a similar taskbar setup, but the icons weren't in one row, but took up the available horizontal and vertical space. 

Is there a way to get the same thing with this mod? 

![image](https://github.com/user-attachments/assets/3426c80c-6b93-46df-b855-da577a93719f)

I know I have a lot of crap running, but I'm using all of it. 
I've already disabled the widgets, so the weather stuff isn't displayed anymore. 

I made photoshop mockup of what I mean:
![taskbar_modified_tray](https://github.com/user-attachments/assets/7cd7a220-1cb9-4920-ab9d-0b0464d9f71a)
</div>

**Comment by m417z** [comment]
Currently that's not possible with the mod. I looked at having a grid of notification icons, but unfortunately, that's not so easy, as the control that's used by the taskbar (Xaml.Controls.StackPanel) is designed for arranging elements in a single line. Working around that requires some research.

One workaround is, of course, to keep some/most tray icons in the overflow menu. Another workaround is to switch the control orientation to horizontal, which is supported by the control. The downside is that the amount of icons that can fit is limited by the taskbar width, but perhaps it'll work for you.

The orientation, and additional styles such as spacing between items, can be customized with the [Windows 11 Taskbar Styler](https://windhawk.net/mods/windows-11-taskbar-styler) mod. Let me know if you want to try it and I'll help configuring it.

![image](https://github.com/user-attachments/assets/bb33eaf1-20ff-4603-a981-d5bb6df108de)

**Comment by RSA-User** [comment]
I also use the Taskbar positioned vertically.  I like the workaround idea of switching the control orientation so the notification icons display in a row instead of a column.  Unfortunately I don't have the tech know-how to figure out what Target and Style statements to use in Windows 11 Taskbar Styler to achieve this.  Any help would be appreciated.

**Comment by m417z** [comment]
First, install the [Windows 11 Taskbar Styler](https://windhawk.net/mods/windows-11-taskbar-styler) mod, then go to the mod's settings and add the following targets and styles under "Control styles":

### Switching the tray icons orientation

Target:
```
SystemTray.NotificationAreaIcons > ItemsPresenter > StackPanel
```
Style:
```
Orientation=0
```

By default, there will be no spacing between the icons. To add spacing, use the target:
```
SystemTray.NotificationAreaIcons > ItemsPresenter > StackPanel > ContentPresenter
```
Style:
```
Margin=0,4,0,4
```
Replace `4` with the desired spacing size.

### Switching the system tray icons orientation

Use the same targets and styles as for the tray icons, but replace:
```
SystemTray.NotificationAreaIcons > ItemsPresenter
```
With:
```
SystemTray.OmniButton#ControlCenterButton > Grid > ContentPresenter > ItemsPresenter
```
in the targets.

**Comment by RSA-User** [comment]
That works great.  Thanks!

**Comment by Jerrk** [comment]
Hopeful for the future 😃

**Comment by scbunker** [comment]
Thanks! Is there a way to change the justification of the icons? Preferably centered, but left would be good, too. (I have my vertical taskbar on the right of my screen so the icons are really close to the edge. I have another screen to the right, so mouse easily slops off to the other screen.)

![Image](https://github.com/user-attachments/assets/fbac8b3a-327d-4c97-944a-e3701e821879)

**Comment by m417z** [comment]
@scbunker you can use the `VerticalAlignment=Center` style. I updated my comment above.

**Comment by scbunker** [comment]
Thanks! Here is what that looks like for anyone interested.

![Image](https://github.com/user-attachments/assets/aa6dff5f-9b05-41fa-9270-0bcd9ed19388)

**Comment by znkrxtw** [comment]
Is there a way to wrap the tray icons? 
![Image](https://github.com/user-attachments/assets/454eafd1-448e-4807-bd0b-4633f15b678d)
Right now I have 6 icons but the last one is hidden on the right side.

**Comment by m417z** [comment]
@znkrxtw currently there's no way to have a grid of icons, that's what this issue is about.

**Comment by eng3** [comment]
The suggested settings allows the systray icons to be horizontal now.  How about the "quick setting tray", the part next to systray that shows wifi, volume, battery

**Comment by m417z** [comment]
[Taskbar tray icon spacing and grid](https://windhawk.net/mods/taskbar-notification-icon-spacing) v1.2 now has the option to have a grid of tray icons, which can be used together with the Vertical Taskbar for Windows 11 mod.

<img width="80" height="310" alt="Image" src="https://github.com/user-attachments/assets/6903bbc5-bf91-4fad-8c44-df52917009be" />

**Comment by WegnerDan** [comment]
Awesome, thank you!

**Comment by m417z** [comment]
@eng3 right, for system icons, you can still use the tip from here: https://github.com/ramensoftware/windhawk-mods/issues/1068#issuecomment-2546812912.

**Comment by eng3** [comment]
> [@eng3](https://github.com/eng3) right, for system icons, you can still use the tip from here: [#1068 (comment)](https://github.com/ramensoftware/windhawk-mods/issues/1068#issuecomment-2546812912).

I had been using that before this new mod.  It also didnt seem to change the system icons

**Comment by m417z** [comment]
It was shown to affect the system icons in this thread. If it doesn't work for you, perhaps you didn't configure it correctly.

**Comment by eng3** [comment]
Ah, I forgot to read the bottom part.

It does work on one of my computers but not on another.

Sometime in the past week, after updating, the comment 1068 mod stopped working (for both trays)

Both are running v1.5. with
{"theme":"","controlStyles[0].target":"SystemTray.OmniButton#ControlCenterButton > Grid > ContentPresenter > ItemsPresenter > StackPanel","controlStyles[0].styles[0]":"Orientation=0","styleConstants[0]":"","resourceVariables[0].variableKey":"","resourceVariables[0].value":"","controlStyles[0].styles[1]":"VerticalAlignment=Center","controlStyles[1].target":"SystemTray.OmniButton#ControlCenterButton > Grid > ContentPresenter > ItemsPresenter > StackPanel > ContentPresenter","controlStyles[1].styles[0]":"Margin=0,4,0,4"}

The other mods seem to work fine.

**Comment by eng3** [comment]
AH, that worked, thanks!

**Comment by dgtwzd1** [comment]
Leaving a related source breadcrumb here in case it helps with tray/system-icon layout work.

Disclosure: I made this standalone/source-included utility:
https://github.com/dgtwzd1/systray-wrap-doubler

It targets the Windows 11 notification-area/system-tray area and makes tray icons use a compact two-row layout beside the clock/date area.

The part that may be useful here is the approach: it mutates the real Explorer XAML tray containers directly rather than hiding/duplicating icons.

Relevant source:
https://github.com/dgtwzd1/systray-wrap-doubler/blob/main/src/TrayHook.Native/TrayHook.Native.cpp

The project is MIT licensed/source-included, so if any part of the tray/system-icon handling is useful for Windhawk mods, feel free to take it.
